kv/kvserver: TestFlowControlSendQueueRangeFeed failed

kv/kvserver.TestFlowControlSendQueueRangeFeed [failed](https://mesolite.cluster.engflow.com/invocations/default/7468ce23-021a-42e1-9907-5d5a7ed97bda?testReportRun=5&testReportShard=36&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [f09626fe58245d5fe78d318cffec5a3664e8fa6e](https://github.com/cockroachdb/cockroach/commits/f09626fe58245d5fe78d318cffec5a3664e8fa6e):


```
=== RUN   TestFlowControlSendQueueRangeFeed
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestFlowControlSendQueueRangeFeed1049906775
    test_log_scope.go:82: use -show-logs to present logs inline
    flow_control_integration_test.go:3413: condition failed to evaluate within 45s: from flow_control_integration_test.go:3269: expected rangefeed on n3, got n1
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestFlowControlSendQueueRangeFeed1049906775
--- FAIL: TestFlowControlSendQueueRangeFeed (45.99s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=36</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #156319 kv/kvserver: TestFlowControlSendQueueRangeFeed failed [C-test-failure O-robot T-kv branch-release-25.4 release-blocker]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFlowControlSendQueueRangeFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-55950

--- a/pkg/kv/kvserver/flow_control_integration_test.go
+++ b/pkg/kv/kvserver/flow_control_integration_test.go
@@ -3359,7 +3359,7 @@ func TestFlowControlSendQueueRangeFeed(t *testing.T) {
 	h.resetV2TokenMetrics(ctx)
 	h.waitForConnectedStreams(ctx, desc.RangeID, 3, 0 /* serverIdx */)
 
-	ts := tc.Server(2)
+	srv2 := tc.Server(2)
 	span := desc.KeySpan().AsRawSpanWithNoLocals()
 	ignoreValues := func(event kvcoord.RangeFeedMessage) {}
 
@@ -3400,17 +3400,26 @@ func TestFlowControlSendQueueRangeFeed(t *testing.T) {
   WHERE name LIKE 'kv.rangefeed.closed_timestamp.slow_ranges.cancelled'
   ORDER BY name ASC;
 `
-
-	closeFeed := rangeFeed(
-		ctx,
-		ts.DistSenderI(),
-		span,
-		tc.Server(0).Clock().Now(),
-		ignoreValues,
-		kvcoord.WithRangeObserver(observer),
-	)
+	// The rangefeed is supposed to live on n3, since that's srv2 and it is
+	// supposed to prefer the local replica. However, it can happen that n3
+	// doesn't see itself in the gossip network yet. Retry until the rangefeed
+	// does get planned on n3.
+	var closeFeed func()
+	testutils.SucceedsSoon(t, func() error {
+		if closeFeed != nil {
+			closeFeed()
+		}
+		closeFeed = rangeFeed(
+			ctx,
+			srv2.DistSenderI(),
+			span,
+			tc.Server(0).Clock().Now(),
+			ignoreValues,
+			kvcoord.WithRangeObserver(observer),
+		)
+		return checkRangeFeedNodeID(3, true /* include */)
+	})
 	defer closeFeed()
-	testutils.SucceedsSoon(t, func() error { return checkRangeFeedNodeID(3, true /* include */) })
 	h.comment(`(Rangefeed on n3)`)
 
 	h.comment(`

